### DISCUSSION TITLE
Is the contents of $return guaranteed for the `post_add_new_link` hook?

### DISCUSSION THREAD

**Discussion by jfix** [YOURLS/YOURLS#3017] [Q&A] [answered]
I'm using the hook `post_add_new_link ` to post information on the newly added short link to a Slack channel.

One of the arguments that gets passed to the functions listening to this hook is `$return`. This object is filled in [here for example](https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-shorturls.php#L110).

I've noticed that while developing locally, I was able to obtain all values (especially the IP address and the date time), but once in production, they are no longer provided. I thought that maybe this was due to the fact that I had `YOURLS_DEBUG` set to `true` in dev but not in production, but looking at the code this should not have any influence on the `$return` variable's contents.

My code is here: https://github.com/jfix/yourls-plugin-slack-notifier/blob/4e859064873773a050fb896e0e8c742e0274960f/plugin.php#L20

I'd be grateful for an idea or pointer why the `ip` and `date` values might be empty.

I base my code on the assumption that `$args` contains the following array:

```
/*
$args contains this array:

Array
(
    [0] => https://www.google.com/
    [1] => theKeyWord
    [2] => Google Search
    [3] => Array
        (
            [url] => Array
                (
                    [keyword] => theKeyWord
                    [url] => https://www.google.com/
                    [title] => Google Search
                    [date] => 2021-07-12 12:48:10
                    [ip] => ::1
                )

            [status] => success
            [message] => https://www.google.com/ added to database
            [title] => Google Search
            [html] => a DOM element, table row to be inserted in the admin table maybe?
            [shorturl] => http://localhost:8080/yourls/theKeyWord
        )
)
 */
```

This is what a Slack notification looks like (note the `unknown` IP address):

<img width="624" alt="Screenshot 2021-08-14 at 00 01 42" src="https://user-images.githubusercontent.com/170240/129422399-77efd140-9978-4b5f-bd0d-19f66845af2e.png">

**Comment by ozh** [comment, marked as answer]
There is most likely something happening on the server side when it tries to [get the IP](https://github.com/YOURLS/YOURLS/blob/94f5db040555329b9f10652918a8ea0ca58339bb/includes/functions-shorturls.php#L44). This isn't a failsafe step.

I'd suggest if you can to var_dump the server headers and see if there's something that looks like an IP there

**Reply by jfix** [reply]
OK, I've looked over the code and noticed that in my (not so recent :( ) version of YOURLS the `$return` argument in the hook `yourls_do_action( 'post_add_new_link', $url, $keyword, $title );` is missing. Adding it solves the issue. I will upgrade my version asap. Thanks for the pointer and sorry for the noise.

**Reply by ozh** [reply]
No problem, pleasure to help, oe.cd has always been high in my heart ;)
